velox/expression/CastExpr-inl.h

```diff
@@ -206,29 +206,97 @@ void CastExpr::applyIntToDecimalCastKernel(
       });
 }
 
-template <typename TInput>
-VectorPtr CastExpr::applyDecimalToDoubleCast(
+template <typename FromNativeType, TypeKind ToKind>
+VectorPtr CastExpr::applyDecimalToFloatCast(
     const SelectivityVector& rows,
     const BaseVector& input,
     exec::EvalCtx& context,
-    const TypePtr& fromType) {
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  using To = typename TypeTraits<ToKind>::NativeType;
+
   VectorPtr result;
-  context.ensureWritable(rows, DOUBLE(), result);
+  context.ensureWritable(rows, toType, result);
   (*result).clearNulls(rows);
-  auto resultBuffer =
-      result->asUnchecked<FlatVector<double>>()->mutableRawValues();
+  auto resultBuffer = result->asUnchecked<FlatVector<To>>()->mutableRawValues();
   const auto precisionScale = getDecimalPrecisionScale(*fromType);
-  const auto simpleInput = input.as<SimpleVector<TInput>>();
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  const auto scaleFactor = DecimalUtil::kPowersOfTen[precisionScale.second];
   applyToSelectedNoThrowLocal(context, rows, result, [&](int row) {
-    auto output = util::Converter<TypeKind::DOUBLE, void, false>::cast(
-        simpleInput->valueAt(row));
-    resultBuffer[row] =
-        output / DecimalUtil::kPowersOfTen[precisionScale.second];
+    auto output =
+        util::Converter<ToKind, void, false>::cast(simpleInput->valueAt(row));
+    resultBuffer[row] = output / scaleFactor;
   });
+  return result;
+}
+
+template <typename FromNativeType, TypeKind ToKind>
+VectorPtr CastExpr::applyDecimalToIntegralCast(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  using To = typename TypeTraits<ToKind>::NativeType;
+
+  VectorPtr result;
+  context.ensureWritable(rows, toType, result);
+  (*result).clearNulls(rows);
+  auto resultBuffer = result->asUnchecked<FlatVector<To>>()->mutableRawValues();
+  const auto precisionScale = getDecimalPrecisionScale(*fromType);
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  const auto scaleFactor = DecimalUtil::kPowersOfTen[precisionScale.second];
+  applyToSelectedNoThrowLocal(context, rows, result, [&](int row) {
+    auto value = simpleInput->valueAt(row);
+    auto integralPart = value / scaleFactor;
+    auto fractionPart = value % scaleFactor;
+    auto sign = value >= 0 ? 1 : -1;
+    bool needsRoundUp =
+        (scaleFactor != 1) && (sign * fractionPart >= (scaleFactor >> 1));
+    integralPart += needsRoundUp ? sign : 0;
+    if (integralPart > std::numeric_limits<To>::max() ||
+        integralPart < std::numeric_limits<To>::min())
+      VELOX_USER_FAIL(makeErrorMessage(input, row, toType) + "Out of bounds.");
```

Comment:
if inside a try_cast, we can set null in the output directly instead of throwing exception here, 
throw+ constructing the exception are expensive, if this happen for each row for some reason it would be 100sX slower.

Reply:
check void CastExpr::applyCastKernel

if (inputRowValue.size() == 0) {
        if (setNullInResultAtError()) {
          result->setNull(row, true);
        } else {
          context.setVeloxExceptionError(
              row,
              makeBadCastException(
                  result->type(), *input, row, "Empty string"));
        }
        return;
      }

Reply:
Thanks @laithsakka. I didn't know there was such mechanism. Changed accordingly.